src/main/perl/RuleBasedEditor.pm

```diff
@@ -0,0 +1,1175 @@
+                                qr/^$INTRO_PATTERN/,
+                                qr/^$intro$/,
+                                $intro . "\n#\n",
+                                BEGINNING_OF_FILE,
+                               );
+
+    $self->_apply_rules(
+                        $config_rules,
+                        $config_options,
+                        $parser_options
+                       );
+
+    return 1;
+}
+
+
+=pod
+
+=back
+
+=head2 Private methods
+
+=over
+
+=item formatAttrValue
+
+This function formats an attribute value based on the value format specified.
+
+Arguments:
+    attr_value : attribue value
+    line_fmt : line format (see LINE_FORMAT_xxx constants)
+    value_fmt : value format (see LINE_VALUE_xxx constants)
+    value_opt : value interpretation/formatting options (bitmask, see LINE_VALUE_OPT_xxx constants)
+
+Return value:
+    A string corresponding to the value formatted according to the format specified by arguments
+    or undef in case of an internal error (missing arguments)
+
+=cut
+
+sub _formatAttributeValue
+{
+    my $function_name = "_formatAttributeValue";
+    my ($self, $attr_value, $line_fmt, $value_fmt, $value_opt) = @_;
+
+    unless (defined($attr_value)) {
+        *$self->{LOG}->error("$function_name: 'attr_value' argument missing (internal error)");
+        return undef;
+    }
+    unless (defined($line_fmt)) {
+        *$self->{LOG}->error("$function_name: 'list_fmt' argument missing (internal error)");
+        return undef;
+    }
+    unless (defined($value_fmt)) {
+        *$self->{LOG}->error("$function_name: 'value_fmt' argument missing (internal error)");
+        return undef;
+    }
+    unless (defined($value_opt)) {
+        *$self->{LOG}->error("$function_name: 'value_opt' argument missing (internal error)");
+        return undef;
+    }
+
+    *$self->{LOG}->debug(2,
+                         "$function_name: formatting attribute value >>>$attr_value<<< (line fmt=$line_fmt, value fmt=$value_fmt, value_opt=$value_opt)"
+                        );
+
+    my $formatted_value;
+    if ($value_fmt == LINE_VALUE_BOOLEAN) {
+        $formatted_value = $attr_value ? 'yes' : 'no';
+
+    } elsif ($value_fmt == LINE_VALUE_INSTANCE_PARAMS) {
+        $formatted_value = '';    # Don't return undef if no matching attributes is found
+                                  # Instance parameters are described in a nlist
+        $formatted_value .= " -l $attr_value->{logFile}"    if $attr_value->{logFile};
+        $formatted_value .= " -c $attr_value->{configFile}" if $attr_value->{configFile};
+        $formatted_value .= " -k $attr_value->{logKeep}"    if $attr_value->{logKeep};
+
+    } elsif ($value_fmt == LINE_VALUE_ARRAY) {
+        *$self->{LOG}->debug(2, "$function_name: array values received: ", join(",", @$attr_value));
+        if ($value_opt & LINE_VALUE_OPT_UNIQUE) {
+            my %values = map(($_ => 1), @$attr_value);
+            $attr_value = [keys(%values)];
```

Comment:
They are sorted because it makes them easier to remove duplicates if you don't guarantee the order and I have not see a use case where the order was important. Could always be changed in the future...